File: src/main/java/org/example/datnbbook/service/TheLoaiService.java
package org.example.datnbbook.service;

import org.example.datnbbook.dto.TheLoaiDTO;
import org.example.datnbbook.model.TheLoai;
import org.example.datnbbook.repository.TheLoaiRepository;
import org.springframework.beans.BeanUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class TheLoaiService {
    @Autowired
    private TheLoaiRepository theLoaiRepository;


    public List<TheLoai> getAll() {
        return theLoaiRepository.findByDeletedFalse();
    }

    public TheLoai getById(Integer id) {
        return theLoaiRepository.findById(id)
                .filter(sp -> !sp.getDeleted())
                .orElseThrow(() -> new RuntimeException("Thể loại không tồn tại với id: " + id));
    }

    public TheLoai create(TheLoai sp) {
        if (sp.getMaTheLoai() == null || sp.getMaTheLoai().isEmpty()) {
            String nextMaTheLoai = theLoaiRepository.getNextSequenceValue();
            sp.setMaTheLoai(nextMaTheLoai);
        }
        sp.setDeleted(false); // Đảm bảo Thể loại mới không bị đánh dấu xóa
        return theLoaiRepository.save(sp);
    }

    public TheLoaiDTO update(Integer id, TheLoaiDTO theLoaiDTO) {
        TheLoai theLoai = theLoaiRepository.findById(id)
                .filter(sp -> !sp.getDeleted())
                .orElseThrow(() -> new RuntimeException("Thể loại không tồn tại với id: " + id));

        BeanUtils.copyProperties(theLoaiDTO, theLoai ,"maTheLoai","id", "deleted");
        theLoai = theLoaiRepository.save(theLoai);
        return convertToDTO(theLoai);
    }

    public void delete(Integer id) {
        TheLoai theLoai = theLoaiRepository.findById(id)
                .filter(sp -> !sp.getDeleted())
                .orElseThrow(() -> new RuntimeException("Thể loại không tồn tại với id: " + id));
        theLoai.setDeleted(true);
        theLoaiRepository.save(theLoai);
    }

    public List<TheLoai> search(String keyword) {
        return theLoaiRepository.search(keyword);
    }

    private TheLoaiDTO convertToDTO(TheLoai theLoai) {
        TheLoaiDTO theLoaiDTO = new TheLoaiDTO();
        BeanUtils.copyProperties(theLoai, theLoaiDTO);
        return theLoaiDTO;
    }
}

File: src/frontend/vue-notus-main/vue-notus-main/vue-notus-main/src/service/TheLoaiService.js
import api from './api';
import TheLoai from '../models/TheLoai';

class TheLoaiService {
  async getAll() {
    const response = await api.get('/the-loai');
    return response.data.map(item => new TheLoai(item));
  }

  async getById(id) {
    const response = await api.get(`/the-loai/${id}`);
    return new TheLoai(response.data);
  }

  async create(theLoai) {
    const response = await api.post('/the-loai', theLoai.toJSON());
    return new TheLoai(response.data);
  }

  async update(id, theLoai) {
    const response = await api.put(`/the-loai/${id}`, theLoai.toJSON());
    return new TheLoai(response.data);
  }

  async delete(id) {
    await api.delete(`/the-loai/${id}`);
  }

  async search(keyword) {
    const response = await api.get('/the-loai/search', {
      params: { keyword }
    });
    return response.data.map(item => new TheLoai(item));
  }
}

export default new TheLoaiService();